File: src/main/java/Grid.java
public class Grid {

    public static final int MAX_WIDTH = 10;
    public static final int MAX_HEIGHT = 10;

    private Direction direction = Direction.NORTH;

    private Point position = new Point(0, 0);

    public void calculatePosition(char command) {
        if (command == 'R') {
            turnRight();
        } else if (command == 'L') {
            turnLeft();
        } else if (command == 'F') {
            moveForward();
        } else if (command == 'B') {
            moveBackward();
        }
    }

    private void moveForward() {
        switch (this.direction) {
            case EAST:
                incrementX();
                break;
            case WEST:
                decrementX();
                break;
            case NORTH:
                decrementY();
                break;
            case SOUTH:
                incrementY();
                break;
        }
    }

    private void moveBackward() {
        switch (this.direction) {
            case EAST:
                decrementX();
                break;
            case WEST:
                incrementX();
                break;
            case NORTH:
                incrementY();
                break;
            case SOUTH:
                decrementY();
                break;
        }
    }

    private void incrementY() {
        int y = (this.position.y() + 1) % MAX_HEIGHT;
        this.position = new Point(this.position.x(), y);
    }

    private void decrementY() {
        int y = this.position.y() == 0 ? 9 : this.position.y() - 1;
        this.position = new Point(this.position.x(), y);
    }

    private void decrementX() {
        int x = this.position.x() == 0 ? 9 : this.position.x() - 1;
        this.position = new Point(x, this.position.y());
    }

    private void incrementX() {
        int x = (this.position.x() + 1) % MAX_WIDTH;
        this.position = new Point(x, this.position.y());
    }

    private void turnLeft() {
        this.direction = direction.left();
    }

    private void turnRight() {
        this.direction = direction.right();
    }

    public String reportPosition() {
        return this.position.x() + ":" + this.position.y() +  ":" + this.direction.value();
    }

}

File: scala/src/main/scala/Grid.scala


class Grid(obstacles: List[Position]) {

  val directions = List(new North(), new East(), new South(), new West())

  def right(actualDirection: Direction): Direction = {
    val actualDirectionIndex = directions.indexOf(directions.find(direction => actualDirection.value == direction.value).head)
    directions((actualDirectionIndex + 1) % 4)
  }

  def left(actualDirection: Direction): Direction = {
    val actualDirectionIndex = directions.indexOf(directions.find(direction => actualDirection.value == direction.value).head)
    if (actualDirectionIndex == 0) {
      directions.last
    } else {
      directions(actualDirectionIndex - 1)
    }
  }

  def obstacleIn(position: Position): Boolean = obstacles.contains(position)

}

object Grid {

  val MAX_WIDTH = 10

  val MAX_HEIGHT = 10

  def incrementX(position: Position): Position = {
    position.copy(x = (position.x + 1) % Grid.MAX_WIDTH)
  }

  def incrementY(position: Position): Position = {
    position.copy(y = (position.y + 1) % Grid.MAX_HEIGHT)
  }

  def decrementX(position: Position): Position = {
    position.copy(x = if (position.x == 0) Grid.MAX_WIDTH - 1 else position.x - 1)
  }

  def decrementY(position: Position): Position = {
    position.copy(y = if (position.y == 0) Grid.MAX_HEIGHT - 1 else position.y - 1)
  }

}

abstract class Direction {

  def value: Char

  def forward(position: Position): Position

  def backward(position: Position): Position

}

class North extends Direction {

  val value = 'N'

  override def forward(position: Position): Position = Grid.decrementY(position)

  override def backward(position: Position): Position = Grid.incrementY(position)

}

class East extends Direction {

  val value = 'E'

  override def forward(position: Position): Position = Grid.incrementX(position)

  override def backward(position: Position): Position = Grid.decrementX(position)

}

class South extends Direction {

  val value = 'S'

  override def forward(position: Position): Position = Grid.incrementY(position)

  override def backward(position: Position): Position = Grid.decrementY(position)

}

class West extends Direction {

  val value = 'W'

  override def forward(position: Position): Position = Grid.decrementX(position)

  override def backward(position: Position): Position = Grid.incrementX(position)

}

